compiler/rustc_mir_build/src/build/expr/as_rvalue.rs

```diff
@@ -393,51 +394,51 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
         this.cfg.push(block, Statement { source_info, kind: StatementKind::StorageLive(temp) });
 
-        let arg_place = unpack!(block = this.as_place(block, arg));
-
-        let mutability = match arg_place.as_ref() {
-            PlaceRef { local, projection: &[] } => this.local_decls[local].mutability,
-            PlaceRef { local, projection: &[ProjectionElem::Deref] } => {
-                debug_assert!(
-                    this.local_decls[local].is_ref_for_guard(),
-                    "Unexpected capture place",
-                );
-                this.local_decls[local].mutability
-            }
-            PlaceRef {
-                local,
-                projection: &[ref proj_base @ .., ProjectionElem::Field(upvar_index, _)],
-            }
-            | PlaceRef {
-                local,
-                projection:
-                    &[ref proj_base @ .., ProjectionElem::Field(upvar_index, _), ProjectionElem::Deref],
-            } => {
-                let place = PlaceRef { local, projection: proj_base };
-
-                // Not projected from the implicit `self` in a closure.
-                debug_assert!(
-                    match place.local_or_deref_local() {
-                        Some(local) => local == Local::new(1),
-                        None => false,
-                    },
-                    "Unexpected capture place"
-                );
-                // Not in a closure
-                debug_assert!(
-                    this.upvar_mutbls.len() > upvar_index.index(),
-                    "Unexpected capture place"
-                );
-                this.upvar_mutbls[upvar_index.index()]
+        let arg_place_builder = unpack!(block = this.as_place_builder(block, arg));
+
+        let mutability = match arg_place_builder.base() {
+            // Capture parent's local variable
+            PlaceBase::Local(local) =>  this.local_decls[local].mutability,
+            // Parent is a closure and we capturing a variable that is captured
+            // by the parent itself.
```

Comment:
I think a comment would be good here-- something like "Extract mutability from the parent"? is that what is going on here, or did I misunderstand :) 

also, this is the code we talked about rewriting a bit, right?

Reply:
Not quite. How we build `upvar_mutbls` will change. I have a commit here which essentially represents that should be needed https://github.com/rust-lang/rust/pull/79696/commits/fae4d2e855413ead921e4634c5e1101c26825ae8

Reply:
Not quite was for the rewriting.

Reply:
Yes as I'll update the comment to talk about mutability but we assume that the mutability for the current place in the nested closure is the same as the mutability of the place in the parent closure.